std-bits/base/src/main/java/org/enso/base/Text_Utils.java

```diff
@@ -253,7 +253,7 @@ public static long char_length(String str) {
    * @param str the string to measure
    * @return length of the string
    */
-  public static long grapheme_length(String str) {
+  private static long grapheme_length(String str) {
```

Comment:
it's a bit annoying that we keep the 2 identical implementations. Can we have them in one place?
Also, since this is now private, we probably don't have to use enso naming of the method with underscores?

Reply:
We could also probably remove this from `Text_Utils.java` quite easily (only used by `starts_with` and `ends_with` now).

Reply:
Is there any code/project shared between Enso engine and standard library? I can put it there to share the Java code. However the engine is built as a fat JAR, isn't it? E.g. the code will anyway be duplicated. It has to be as engine JAR is loaded by a different classloader than std-lib. Btw. in the [native-image+espresso setup](https://www.pivotaltracker.com/story/show/183260380) the engine would be running in `native-image` and the libraries in Espresso - e.g. we cannot avoid duplication in the runtime.

> (only used by starts_with and ends_with now).

Right. As such I kept the method there, `private`.

Now advice: what am I suppose to do? Ignore DRY? Rewrite `start_with` & co. somehow?

Reply:
Right. So how about having utils project that can be shared between engine and std-bits? It can't be the first time we have this kind of duplication. Or is that too much of an overkill?
If it can't be done easily atm, then it can't be done, not gonna block this PR. But it would be nice if we at least explored the option ;)

Reply:
@hubertp , do you want to prepare a patch for `build.sbt` to do that? I can then move the code.

Btw. the bigger problem is that `com.ibm.icu:icu4j` is there twice, in my opinion! We shouldn't need `icu4j` in the engine, but the dependency is already there.

The new project will have to depend on this library and then the huge `icu4j` library needs to be copied at both ends - engine as well as std-lib. I'd say you opened a _can of worms_...

Reply:
I guess I can let that one slip...

Reply:
I don't want to resurrect this thread, but given that all of us don't like this repetition too much _and_ it's quite non-trivial to extract the shared code (although I think this is something that we will need to do at some point anyway) _and_ that it is used by just these two mentioned functions _and_ the fact that these functions not necessarily need to be written that way exactly.
So maybe we can try to rewrite them to just not use this function?

Reply:
Added a chore ticket to do this: https://www.pivotaltracker.com/story/show/183794867

Reply:
Perfect, thanks!